File: implementations/python/tools.py
import json
import os
from registry import registry

# Load the mock tool schemas from spec
SPEC_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "spec"))
with open(os.path.join(SPEC_DIR, "mock_tools.json"), "r") as f:
    mock_tools_spec = json.load(f)

# Find specific schema by name
def get_schema(name: str):
    for tool in mock_tools_spec:
        if tool["name"] == name:
            return tool["parameters"]
    return {}

# 1. Weather Tool
def get_weather(location: str, unit: str = "celsius") -> str:
    """Mock implementation of get_weather."""
    # In a real app, this would call a weather API.
    if "san francisco" in location.lower():
        return f"The weather in {location} is 16 degrees {unit.capitalize()} and foggy."
    return f"The weather in {location} is 22 degrees {unit.capitalize()} and sunny."

registry.register(
    name="get_weather",
    description="Get the current weather in a given location.",
    func=get_weather,
    schema=get_schema("get_weather")
)

# 2. Read File Tool
def read_file(path: str) -> str:
    """Mock implementation of read_file."""
    try:
        with open(path, "r", encoding="utf-8") as f:
            return f.read()
    except Exception as e:
        return f"Error reading file: {str(e)}"

registry.register(
    name="read_file",
    description="Read the contents of a file.",
    func=read_file,
    schema=get_schema("read_file")
)

File: implementations/typescript/src/tools.ts
import * as fs from 'fs';
import * as path from 'path';
import { registry } from './ToolRegistry';

// Load mock tool schemas from spec
const specPath = path.resolve(__dirname, '../../../spec/mock_tools.json');
const mockToolsSpec = JSON.parse(fs.readFileSync(specPath, 'utf8'));

function getSchema(name: string): any {
  const tool = mockToolsSpec.find((t: any) => t.name === name);
  return tool ? tool.parameters : {};
}

// 1. Weather Tool
registry.register(
  'get_weather',
  'Get the current weather in a given location.',
  async (args: { location: string; unit?: string }) => {
    const location = args.location || '';
    const unit = args.unit || 'celsius';
    
    if (location.toLowerCase().includes('san francisco')) {
      return `The weather in ${location} is 16 degrees ${unit} and foggy.`;
    }
    return `The weather in ${location} is 22 degrees ${unit} and sunny.`;
  },
  getSchema('get_weather')
);

// 2. Read File Tool
registry.register(
  'read_file',
  'Read the contents of a file.',
  async (args: { path: string }) => {
    try {
      return fs.readFileSync(args.path, 'utf8');
    } catch (e: any) {
      return `Error reading file: ${e.message}`;
    }
  },
  getSchema('read_file')
);
